std/json.c3

```diff
@@ -0,0 +1,428 @@
+// function specializations.
+function new_json_node : (node_type : int) -> JSON_Node = {
+    return {
+        node_type,
+        make_optional<StringView>(),
+        make_optional<Vector<JSON_Node>>(),
+    };
+}
+
+function char : (literal : u8[1]&) -> u8 = {
+    return literal[0];
+}
+
+function is_eof : (cursor : Cursor&) -> bool = {
+    return cursor.position >= cursor.data.length;
+}
+
+function peak : (cursor : Cursor&) -> Optional<u8> = {
+    if !cursor:is_eof() {
+        return make_optional(cursor.data.data[cursor.position]);
+    }
+
+    return make_optional<u8>();
+}
+
+function next : (cursor : Cursor&) -> Optional<u8> = {
+    // TODO this should be const (can't use with has_value() right now).
+    let token : Optional<u8> = cursor:peak();
+
+    if token:has_value() {
+        cursor.position += 1;
+    }
+
+    return token;
+}
+
+function is_json_whitespace : (token : u8) -> bool = {
+    // JSON defines whitespace as one of { space, linefeed, carriage return,
+    // horizontal tab }. If we use string literals instead of code points, then
+    // our code doesn't care about the underlying encoding :).
+    return token == char(" ") or token == char("\n")
+           or token == char("\r") or token == char("\t");
+}
+
+function eat : (cursor : Cursor&, token : u8) -> bool = {
+    // TODO this should be const (can't use with has_value() right now).
+    let next_char : Optional<u8> = cursor:peak();
+
+    let next_char_data : u8 = next_char:data();
+    if next_char:has_value() and next_char_data == token {
+        cursor:next();
+        return true;
+    }
+
+    return false;
+}
+
+function eat : (cursor : Cursor&, string : StringView) -> bool = {
+    for token in string:get_iter() {
+        if !cursor:eat(token) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+function eat_whitespace : (cursor: Cursor&) -> void = {
+    while !cursor:is_eof() {
+        let next_char : Optional<u8> = cursor:peak();
+        runtime_assert(next_char:has_value());
+
+        if !is_json_whitespace(next_char:data()) {
+            return;
+        }
+
+        // Eat the current character.
+        cursor:next();
+    }
+}
+
+// TODO return ErrorOr<JSON_Node>.
+function parse_object : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    if !cursor:eat(char("{")) {
+        return make_optional<JSON_Node>();
+    }
+
+    let node : JSON_Node = new_json_node(/* object */ 6);
+
+    cursor:eat_whitespace();
+
+    if cursor:eat(char("}")) {
+        // Empty object.
+        return make_optional(node);
+    }
+
+    // Initialize children vector.
+    node.children:store(cursor.allocator:make_vector<JSON_Node>());
+
+    while true {
+        cursor:eat_whitespace();
+
+        let string_node : Optional<JSON_Node> = cursor:parse_string();
+
+        if !string_node:has_value() {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
```

Comment:
This is good for now but once we have `map` (or similar) it'd be nice to express that the key can only be a string using the type-system. Then it could just be map<StringView, JSON_Node>

Reply:
That'd be ideal, although I'd rather go with some sort of tree vs a hash map to avoid allocating a bunch of buckets for three elements